chainer/functions/array/spatial_transformer_sampler.py

```diff
@@ -0,0 +1,288 @@
+            gvs.append(gv_tmp)
+        gu = xp.concatenate([xp.expand_dims(g, axis=0) for g in gus], axis=0)
+        gv = xp.concatenate([xp.expand_dims(g, axis=0) for g in gvs], axis=0)
+
+        gu *= gy
+        gv *= gy
+        gu = xp.sum(gu, axis=1)
+        gv = xp.sum(gv, axis=1)
+        # this offsets scaling of the coordinates
+        gu = gu / 2. * (W - 1)
+        gv = gv / 2. * (H - 1)
+
+        ggrid = xp.concatenate((gu[:, None], gv[:, None]), axis=1)
+
+        # --- gx
+        gxs = []
+        samples_arange = xp.arange(out_H * out_W, dtype=numpy.int32)
+        for b in range(B):
+            dydx = xp.zeros((H, W, out_H * out_W), dtype=gy.dtype)
+            if xp is numpy:
+                scatter_add = numpy.add.at
+            else:
+                scatter_add = xp.scatter_add
+            scatter_add(dydx, (v0[b], u0[b], samples_arange), wu1[b] * wv1[b])
+            scatter_add(dydx, (v0[b], u1[b], samples_arange), wu0[b] * wv1[b])
+            scatter_add(dydx, (v1[b], u0[b], samples_arange), wu1[b] * wv0[b])
+            scatter_add(dydx, (v1[b], u1[b], samples_arange), wu0[b] * wv0[b])
+
+            gy_tmp = gy[b].reshape(C, -1)
+            gy_tmp = gy_tmp.transpose(1, 0)
+            gx = dydx.dot(gy_tmp)
+            gxs.append(gx.transpose(2, 0, 1))
+        gx = xp.concatenate([xp.expand_dims(g, axis=0) for g in gxs], axis=0)
+        return gx, ggrid
+
+
+def spatial_transformer_sampler(x, grid, use_cudnn=True):
+    """2D Spatial Transformer sampler.
+
+    This is a differentiable image sampler. With a set of sampling points
+    ``grid`` and an input feature map ``x``, this produces a sampled output
+    feature map.
+
+    This function currently only supports bilinear interpolation as a sampling
+    kernel.
+
+    It is important to note that this function assumes values in ``grid`` to
+    be in range :math:`[-1, 1]`.
+
+    Notatition: here is a notation for dimensionalities.
+
+    - :math:`n` is the batch size.
+    - :math:`c_I` is the number of the input channels.
+    - :math:`h` and :math:`w` are the height and width of the input image,
+      respectively.
+    - :math:`h_O` and :math:`w_O` are the height and width of the output
+      image.
+
+    See detail in a paper: `Spatial Transformer Networks \
+    <https://arxiv.org/abs/1506.02025>`_.
+
+    Args:
+        x (~chainer.Variable):  Input variable of shape :math:`(n, c_I, h, w)`.
+        grid (~chainer.Variable): Coordinate variable of shape
+            :math:`(n, 2, h_O, w_O)`. Each coordinate defines the spatial
+            location in the input where a sampling kernel is applied to get
+            the value at a particular pixel in the output.
+            ``grid[idx, :, i, j]`` corresponds to the coordinate that is used
+            to sample the values for an output pixel at location
+            :math:`(i, j)`.
+
+            In the 2nd dimension, the first coordinate corresponds to the
```

Comment:
It might be worth noting that the coordinate (-1, -1) corresponds to the upper-left corner of the input image.